Refactor LiveText to use diff-match-patch

codeofdusk opened the pull request:
### Link to issue number:
Closes #3200. Closes #11498.

### Summary of the issue:
`LiveText` objects use [difflib](https://docs.python.org/3/library/difflib.html) to calculate text changes:

* Difflib is slow, making 10,000-line objects completely unusable and slowing down or locking up NVDA for other large objects.
* Comparisons are made at the line level, so edits in the middle of lines cause characters to the right of the caret to be read out.
* In 21H1 UIA console, when a large amount of text is written, the diff lags far behind the actual console's output, making autoread unusable until NVDA or the console is restarted.

### Description of how this pull request fixes the issue:
This PR replaces difflib with [diff-match-patch](https://pypi.org/project/diff-match-patch/) (DMP) for `LiveText` objects, and changes diffing functions to operate at the string rather than line level.

### Testing performed:
* Tested NVDA in an unbounded (10,000-line) UIA console. This caused a freeze with Difflib, but with DMP the console is usable although somewhat laggy.
* Tested deleting and inserting characters in the middle of a line both with difflib and DMP. With difflib the characters to the right of the caret were read out (as stated in #3200). With DMP only the changed characters were read.
* Re-tested the steps from microsoft/terminal#5481 and confirmed that they are no longer reproducible.
* Installed a [signed build](https://ci.appveyor.com/api/buildjobs/49v47i43kyb5ufhu/artifacts/output%2Fnvda_snapshot_try-livetext-dmp_11500-20753%2C096ad76c.exe) of this PR and ran a number of Windows and Linux console applications as part of daily use.

### Known issues with pull request:
* While this PR really helps in many extreme cases, there are others where it is actually slower than difflib by about half a second. I propose a follow-up PR to switch to [the C++ version](https://github.com/leutloff/diff-match-patch-cpp-stl), either implemented as [a Python extension](https://github.com/JoshData/diff_match_patch-python) or as part of `NVDAHelper`. Initial testing suggests that the C++ library is much faster in these cases.
* Since this PR alone represents a fairly major change, I'd like to get it into 2020.4 fairly early (to get lots of testing time in on `master`).

### Change log entry:
== Bug Fixes ==
- In terminal programs, when inserting or deleting characters in the middle of a line, the characters to the right of the caret are no longer read out. (#3200)

== Changes For Developers ==
- `LiveText._getTextLines` is deprecated. Now override _getText, which returns a string of all text in the object. (#11500)
- `LiveText` objects now calculate diffs by character. Set `_supportsDMP` to `False` to revert to the previous line-based diffing algorithm. (#11500)

codeofdusk committed:
Refactor LiveText to use diff-match-patch.

codeofdusk commented:
Cc @camlorn @feerrenrut @jcsteh @leonardder @michaelDCurran.

lukaszgo1 commented on source/NVDAObjects/window/winConsole.py:
```diff
@@ -1,8 +1,8 @@
-#This file is covered by the GNU General Public License.
-#See the file COPYING for more details.
-#Copyright (C) 2007-2019 NV Access Limited, Bill Dengler
+# NVDAObjects/WinConsole.py
```
Please remove this line.

codeofdusk replied:
Why? `NVDAObjects/behaviors.py` has a similar one as well...

lukaszgo1 replied:
Please take a look at the [copyright headers page](https://github.com/nvaccess/nvda/wiki/Copyright-headers) especially this part is relevant:

> Note: We decided to remove the filename comment from the top of the file, since it doesn't really add anything, and is a source of error on file rename / copying copyright headers between files.

lukaszgo1 commented on source/winConsoleHandler.py:
```diff
@@ -1,8 +1,8 @@
-#This file is covered by the GNU General Public License.
-#See the file COPYING for more details.
-#Copyright (C) 2009-2018 NV Access Limited, Babbage B.V.
+# winConsoleHandler.py
```
And this as well.

codeofdusk committed:
Revert to line-based diffing for this PR as there are some issues with the character-based approach.

codeofdusk commented:
I was running into some strange bugs with the character-based diffing approaches. I've reduced the scope of this PR to use DMP in a line-based mode (with an updated version of the old algorithm to read partial line changes). This should minimize the potential for breaking changes while still providing significant performance improvements (for all `LiveText` implementations) and usability improvements (for 21H1 UIA console). We should definitely explore character-based diffing in a follow-up though, as this would resolve #3200. The description has been updated to reflect these changes.

codeofdusk committed:
Fix

codeofdusk committed:
Review action.

codeofdusk committed:
Add DMP as a submodule.

leonardder commented on source/NVDAObjects/behaviors.py:
```diff
@@ -263,15 +263,18 @@ def event_textChange(self):
-		@return: The current lines of text.
-		@rtype: list of str
+		@return: The current text ob the object.
+		@rtype: str
```
Please use a type annotation on the method's def

leonardder commented on source/NVDAObjects/behaviors.py:
```diff
@@ -307,59 +310,61 @@ def _monitor(self):
 
-	def _calculateNewText(self, newLines, oldLines):
-		outLines = []
+	def _calculateNewText(self, newText, oldText):
```
```suggestion
	def _calculateNewText(self, newText: str, oldText: str) -> List[str]:
```

LeonarddeR reviewed (commented):
Make sure to update the dependencies seciton of the readme

codeofdusk committed:
Switch back to character diffs and fix several assumptions.

codeofdusk committed:
Use line diffing (difflib) in non-enhanced legacy consoles to report changes in the middle of lines.

codeofdusk committed:
Allow DMP to be disabled in advanced settings in case of compatibility issues and to ease testing.

codeofdusk committed:
Review actions.

codeofdusk committed:
Fixes.

codeofdusk committed:
Clarify naming

leonardder commented:
@codeofdusk there have been several changes in this pr. Does the description still match the current situation of the code?

codeofdusk commented:
Yes – I've been keeping the description up-to-date as I've made changes.

codeofdusk committed:
Update licence to specify that DMP is under Apache.

This should make it compatible with NVDA, otherwise Espeak wouldn't be and it's a fairly major component (it's under GPL3 and has a similar patent clause).

codeofdusk commented:
I'm closing the PR in its current form due to licence incompatibility. I'll open a new one that calls DMP in a separate process to avoid linking with NVDA.
